src/residuals.cc

```diff
@@ -0,0 +1,155 @@
+#include <assert.h>
+#include <algorithm>
+#include <kstring.h>
+#include <vcf.h>
+#include "residuals.h"
+
+using namespace std;
+
+namespace GLnexus {
+
+// convert a BCF record into a string
+static shared_ptr<string> bcf1_to_string(const bcf_hdr_t *hdr, const bcf1_t *bcf) {
+    kstring_t kstr;
+    memset((void*)&kstr, 0, sizeof(kstring_t));
+    vcf_format(hdr, bcf, &kstr);
+    auto retval = make_shared<string>(kstr.s, kstr.l);
+    if (retval->length() > 0)
+        retval->erase(retval->find_last_not_of(" \n\r\t")+1); // get rid of extra spaces at the end
+
+    // cleanup
+    if (kstr.s != NULL)
+        free(kstr.s);
+
+    return retval;
+}
```

Comment:
Minor: with the return value optimization and rvalues I think there's little advantage to returning a `shared_ptr<string>` instead of a `string` -- the latter won't lead to unnecessary copying I think?